File: backend/schemas/subscription.py
from datetime import datetime
from decimal import Decimal
from typing import Literal, Optional

from pydantic import BaseModel, Field

Plan = Literal["basic", "pro", "enterprise"]
SubscriptionStatus = Literal["trial", "active", "past_due", "suspended", "cancelled"]
BillingCycle = Literal["monthly", "yearly"]


class SubscriptionCreate(BaseModel):
    company_id:    int
    plan:          Plan = "basic"
    status:        SubscriptionStatus = "active"
    billing_cycle: BillingCycle = "yearly"
    starts_at:     Optional[datetime] = None
    ends_at:       Optional[datetime] = None
    unit_price:    Decimal = Field(Decimal("0"), ge=0)
    currency:      Optional[str] = Field(None, min_length=3, max_length=3)   # None -> company currency
    tax_rate:      Decimal = Field(Decimal("18.00"), ge=0, le=100)
    features:      Optional[dict] = None
    notes:         Optional[str] = None
    # Convenience: grant this many company-wide licences on creation.
    initial_licenses: int = Field(0, ge=0, le=100)


class SubscriptionUpdate(BaseModel):
    plan:          Optional[Plan] = None
    status:        Optional[SubscriptionStatus] = None
    billing_cycle: Optional[BillingCycle] = None
    starts_at:     Optional[datetime] = None
    ends_at:       Optional[datetime] = None
    unit_price:    Optional[Decimal] = Field(None, ge=0)
    currency:      Optional[str] = Field(None, min_length=3, max_length=3)
    tax_rate:      Optional[Decimal] = Field(None, ge=0, le=100)
    features:      Optional[dict] = None
    notes:         Optional[str] = None


class LicenseGrant(BaseModel):
    subscription_id: int
    site_id:         Optional[int] = None
    quantity:        int = Field(1, ge=1, le=100)
    max_users:       Optional[int] = Field(26, ge=1)
    max_admins:      int = Field(1, ge=0)
    unlimited:       bool = False                        # True -> max_users stored as NULL


class LicenseRevoke(BaseModel):
    reason: Optional[str] = None


class LicenseResponse(BaseModel):
    id:              int
    subscription_id: int
    company_id:      int
    site_id:         Optional[int] = None
    site_name:       Optional[str] = None
    license_key:     str
    status:          str
    max_users:       Optional[int] = None
    max_admins:      int
    granted_by:      Optional[int] = None
    granted_at:      Optional[datetime] = None
    revoked_by:      Optional[int] = None
    revoked_at:      Optional[datetime] = None
    revoke_reason:   Optional[str] = None
    created_at:      Optional[datetime] = None

    model_config = {"from_attributes": True}


class SubscriptionResponse(BaseModel):
    id:            int
    company_id:    int
    company_name:  Optional[str] = None
    plan:          str
    status:        str
    billing_cycle: str
    starts_at:     Optional[datetime] = None
    ends_at:       Optional[datetime] = None
    unit_price:    Decimal
    currency:      str
    tax_rate:      Decimal
    features:      Optional[dict] = None
    notes:         Optional[str] = None
    created_at:    Optional[datetime] = None
    updated_at:    Optional[datetime] = None

    # Derived
    capacity:        Optional[int] = None     # None = unlimited
    admin_cap:       int = 0
    seats_used:      int = 0
    admins_used:     int = 0
    active_licenses: int = 0
    is_valid:        bool = True
    reason_code:     Optional[str] = None
    reason:          Optional[str] = None

    model_config = {"from_attributes": True}


class MySubscriptionResponse(SubscriptionResponse):
    licenses:        list[LicenseResponse] = []
    days_to_renewal: Optional[int] = None

File: frontend/src/types/subscription.ts
export type Plan = "basic" | "pro" | "enterprise";
export type SubscriptionStatus = "trial" | "active" | "past_due" | "suspended" | "cancelled";
export type BillingCycle = "monthly" | "yearly";

export interface License {
  id: number;
  subscription_id: number;
  company_id: number;
  site_id: number | null;
  site_name: string | null;
  license_key: string;
  status: "active" | "revoked";
  max_users: number | null;      // null = unlimited
  max_admins: number;
  granted_by: number | null;
  granted_at: string | null;
  revoked_by: number | null;
  revoked_at: string | null;
  revoke_reason: string | null;
  created_at: string | null;
}

export interface Subscription {
  id: number;
  company_id: number;
  company_name: string | null;
  plan: Plan;
  status: SubscriptionStatus;
  billing_cycle: BillingCycle;
  starts_at: string | null;
  ends_at: string | null;
  unit_price: string;            // Decimal serialised as string
  currency: string;
  tax_rate: string;
  features: Record<string, boolean> | null;
  notes: string | null;
  created_at?: string | null;
  updated_at?: string | null;
  // derived
  capacity: number | null;       // null = unlimited
  admin_cap: number;
  seats_used: number;
  admins_used: number;
  active_licenses: number;
  is_valid: boolean;
  reason_code: string | null;
  reason: string | null;
}

export interface SubscriptionDetail extends Subscription {
  licenses: License[];
  days_to_renewal: number | null;
}

export interface SubscriptionCreate {
  company_id: number;
  plan: Plan;
  status: SubscriptionStatus;
  billing_cycle: BillingCycle;
  starts_at?: string | null;
  ends_at?: string | null;
  unit_price: string;
  currency?: string | null;
  tax_rate: string;
  notes?: string | null;
  initial_licenses: number;
}

export interface SubscriptionUpdate {
  plan?: Plan;
  status?: SubscriptionStatus;
  billing_cycle?: BillingCycle;
  starts_at?: string | null;
  ends_at?: string | null;
  unit_price?: string;
  currency?: string;
  tax_rate?: string;
  notes?: string | null;
}

export interface LicenseGrant {
  subscription_id: number;
  site_id?: number | null;
  quantity: number;
  max_users?: number | null;
  max_admins: number;
  unlimited: boolean;
}
